### PR TITLE
Core `fnr-text` improvements

### PR DESCRIPTION
**Pull request by theNerd247**
Upon reviewing the code for `fnr-text` I've noted many improvements that could be made. 

TL;DR: the `fnr-text` needs to be refactored into 2 layers: A core module that is well-typed and is safe to make all the assumptions that are necessary. A second module that is the user interface for the library that provides a user-friendly interface.

A few first highlevel observations are the following:
  - an architecture diagram for `findr` needs to be constructed (if one already exists then it needs to be included in the typedoc)
  - The types defined need to be revisited. There are many refinements that need to be made (name changes, type refinement, and EDSL construction) to make this library production ready.
  - Fully lean into the typing system that TS provides. Assume users of your library are going to use TS and if they chose not to then they are need to know they are knowingly going into "the danger zone".  This leads to the next suggestion.
  - Implement the idea of "parse don't validate" to avoid needing excessive `if` statements and type-guarding.

### CODE REVIEW

**Comment by theNerd247** [comment]
In regards to the comments about types above I suggest using a runtime type-checking system like [this](https://www.npmjs.com/package/typescript-runtime-types/v/1.0.0). I'm not sure which one is the "best" but it would remove the need to do this manually - and would improve API and code quality in the future.

**Comment by theNerd247** [comment]
In regards to refining the types I've found a [pattern-matching library](https://github.com/gvergnaud/ts-pattern). Coming from an FP background pattern-matching signifigantly simplifies code and reduces errors

**Comment by theNerd247** [comment]
This PR is getting to big. I'm splitting it up into:

- #35 
- #34
